README.md

```diff
@@ -18,25 +20,185 @@ $ npm install --save pipage
 
 ## Usage
 
+For detailed API documentation, see [doc/](doc/)
+
 ```js
-var Pipeline = require( 'pipage' )
+var Pipage = require('pipage')
+```
+
+The pipeline is essentially a duplex stream with internal streams, which can be **added to**, **removed from** and **spliced** at runtime:
+
+```
+readable -> pipage[ transform, transform, ... ] -> writable
 ```
 
+### Creating a pipeline
+
 ```js
-var checksum = new ChecksumStream( metadata )
-var slice = new SliceStream( metadata )
+// A blank pass-through pipeline:
+var pipeline = new Pipage()
+
+// Initialized with already existing streams:
+var pipeline = new Pipage([ transform1, transform2, ... ])
+```
+
+### Adding streams
 
-var pipeline = new Pipeline([ slice, checksum ])
+```js
+// Streams can be appended,
+pipeline.append( stream )
+pipeline.append( stream1, stream2, ..., streamN )
+
+// prepended,
+pipeline.prepend( stream1, stream2, ..., streamN )
+
+// inserted at a specific index,
+pipeline.insert( 3, stream1, ..., streamN )
+
+// which is the same as splicing in streams (add N streams at index 3):
+pipeline.splice( 3, 0, stream1, ..., streamN )
+```
+
+### Removing streams
+
+```js
+// Streams can be shifted off the beginning,
+var firstStream = pipeline.shift()
 
-if( bmap ) {
-  pipeline.splice( pipeline.indexOf( checksum ), 1 )
-  pipeline.append( new BlockMap.FilterStream( bmap ) )
+// or popped off the end,
+var lastStream = pipeline.pop()
+
+// spliced out at a specific index (remove 3 from index 2),
+var removedStreams = pipeline.splice( 2, 3 )
+
+// or removed by reference:
+pipeline.remove( stream )
+```
+
+### Selecting streams
+
+```js
+// Get a stream at a specific index in the pipeline:
+var stream = pipeline.get( 2 )
+var lastStream = pipeline.get( -1 )
+
+// Find a stream in the pipeline:
+var index = pipeline.indexOf( stream )
+var lastIndex = pipeline.lastIndexOf( stream )
+```
+
+### Example
+
+```js
+var Pipage = require('pipage')
+var path = require('path')
+var zlib = require('zlib')
+var unbzip2 = require('unbzip2-stream')
+var xz = require('xz')
+
+var pipeline = new Pipage()
+
+switch( path.extname( filename ) ) {
+  case '.gz':
+    pipeline.prepend( zlib.createUnzip() );
+    break
+  case '.bz':
+  case '.bz2':
+    pipeline.prepend( unbzip2() );
+    break
+  case '.xz':
+    pipeline.prepend( new xz.Decompressor() )
+    break
 }
 
-new BlockReadStream( '/path/to/resin-os.img' )
+fs.createReadStream( filename )
   .pipe( pipeline )
-  .pipe( new BlockWriteStream( '/dev/rdisk2' ) )
-  .once( 'finish', function() {
-    // happy face
-  })
+  .pipe( fs.createWriteStream( destination ) )
+```
+
+### Events
+
+#### Error handling
```

Comment:
> I wonder if this README (or the API docs) should mention that 'error' is the only event that's automatically bound to all streams in the pipeline?

Technically that's not the only event listened to internally (there's also `end`, `readable`, etc. - all the core stream's events that need handling).
That the `error` event is automatically handled should be mentioned though. I'll add that in a bit.

> I guess theoretically there's nothing stopping somebody from unbinding (or rebinding) the error event from a particular stream?

The `error` event cannot be unbound, as the pipeline keeps track of which errors have been bound through `.bind()`, and the `error` event is not one of them, as it's internal and used for proper error propagation.